sdk/identity/azure-identity/azure/identity/_internal/msal_credentials.py

```diff
@@ -57,21 +94,51 @@ def _get_app(self):
         # type: () -> msal.ClientApplication
         pass
 
-    def _create_app(self, cls):
-        # type: (Type[msal.ClientApplication]) -> msal.ClientApplication
+    def _create_app(self, cls, **kwargs):
+        # type: (Type[msal.ClientApplication], **Any) -> msal.ClientApplication
         """Creates an MSAL application, patching msal.authority to use an azure-core pipeline during tenant discovery"""
 
         # MSAL application initializers use msal.authority to send AAD tenant discovery requests
         with self._adapter:
             # MSAL's "authority" is a URL e.g. https://login.microsoftonline.com/common
-            app = cls(client_id=self._client_id, client_credential=self._client_credential, authority=self._base_url)
+            app = cls(
+                client_id=self._client_id,
+                client_credential=self._client_credential,
+                authority=self._base_url,
+                token_cache=self._cache,
+                **kwargs
+            )
 
         # monkeypatch the app to replace requests.Session with MsalTransportAdapter
         app.client.session.close()
         app.client.session = self._adapter
 
         return app
 
+    def _account_matches_profile(self, account):
+        return (
+            account.get("home_account_id") == self._profile.home_account_id
+            and account.get("environment") == self._profile.environment
+        )
```

Comment:
>) [](start = 8, length = 1)

sorry that I was wrong, the home_account_id is the home_tenent_id, we still need and account.get("realm") == self._profile.tenant_id

Reply:
In practical terms I think home_account_id alone is sufficient because it's the only detail of the account MSAL uses to find a token: [finding an access token](https://github.com/AzureAD/microsoft-authentication-library-for-python/blob/7856e8a92dc07e3462488bb09b6fab40c3babb8c/msal/application.py#L547-L552), [finding a refresh token](https://github.com/AzureAD/microsoft-authentication-library-for-python/blob/7856e8a92dc07e3462488bb09b6fab40c3babb8c/msal/application.py#L577-L581). The environment and realm (tenant) are set when the credential's `msal.PublicClientApplication` instance is created. So it looks like this filter could be much simpler: take the first account with a matching home_account_id.

I want to be sure my understanding is correct though. Does the above make sense to you?

Reply:
Yes, since you will loop the accounts (may have multiple tenant accounts per user account) to try to acquire access token, I think this behavior is correct.

---
In reply to: [398946277](https://github.com/chlowell/azure-sdk-for-python/pull/4#discussion_r398946277) [](ancestors = 398946277)